base/MAPL_VerticalMethods.F90

```diff
@@ -369,10 +381,156 @@ subroutine regrid_eta_to_pressure(this,ptrin,ptrout,rc)
+           integer :: p_in_lb, p_in_ub
+           integer :: p_out_lb, p_out_ub
+           integer :: k_bottom, ii
+           real  :: ps_in_mb, hgt, tstar, t0, alph, alnp, tplat, tprime0
+
+           d_in_lb = lbound(data_in,1)
+           d_in_ub = ubound(data_in,1)
+           d_out_lb = lbound(data_out,1)
+           d_out_ub = ubound(data_out,1)
+           p_in_lb = lbound(plevs_in,1)
+           p_in_ub = ubound(plevs_in,1)
+           p_out_lb = lbound(plevs_out,1)
+           p_out_ub = ubound(plevs_out,1)
+
+           if (any(plevs_out > plevs_in(p_in_ub)*0.01)) then
+              k_bottom = -1
+              do ii=p_out_lb,p_out_ub
+                 if (ks(ii) /= -1) then
+                    k_bottom = ii
+                    exit
+                 end if
+              enddo
+              if (k_bottom == -1) k_bottom = size(plevs_out)
+              select case(var_flag)
+              case(VFLAG_GENERIC)
+                 do ii=1,k_bottom
+                    data_out(ii)=data_in(ks(k_bottom))
+                 enddo
+              case(VFLAG_GEOZ)
+                 ps_in_mb = ps*0.01
+                 hgt = phis_ij/MAPL_GRAV
+                 tstar = ts_ij*(1.0+alpha*(ps_in_mb/(plevs_in(d_in_ub)*0.01)-1.0))
+                 t0 = tstar + 0.0065*hgt 
+
+                 if (tstar <= 290.5 .and. t0 > 290.5) then
+                    alph = MAPL_RDRY/phis_ij*(290.5-tstar)
+                 else if (tstar > 290.5 .and. t0 > 290.5) then
+                    alph = 0
+                    tstar = 0.5*(290.5 + tstar)
+                 else
+                    alph = alpha
+                 end if
+                 
+                 if (tstar < 255.0) then
+                    tstar = 0.5*(tstar+255.0)
+                 end if
+                 do ii=1,k_bottom
+                    alnp = alph*log(plevs_out(ii)/ps_in_mb)
+                    data_out(ii) = hgt - MAPL_RDRY*tstar*(1.0/MAPL_GRAV)*log(plevs_out(ii)/ps_in_mb)*(1.0+ 0.5*alnp+1.0/6.0*alnp**2)
+                 enddo
+              case(VFLAG_GEOT)
+                 ps_in_mb = ps*0.01
+                 tstar = data_in(d_in_ub)*(1.0+alpha*(ps_in_mb/(plevs_in(p_in_ub)*0.01)-1.0))
+                 hgt = phis_ij/MAPL_GRAV
+                 do ii=1,k_bottom
+                    if (hgt < 2000.0) then
+                       alnp = alpha*log(plevs_out(ii)/ps_in_mb)
+                    else
+                       t0 = tstar +0.0065*hgt
+                       tplat = min(t0,298.0)
+                       if (hgt <= 2500.0) then
+                          tprime0 = 0.002*((2500.0-hgt)*t0+(hgt-2000.0)*tplat)
+                       else
+                          tprime0 = tplat
+                       end if
+                       if (tprime0 < tstar) then
+                          alnp = 0.0
+                       else 
+                          alnp = MAPL_RDRY*(tprime0-tstar)/phis_ij*log(plevs_out(ii)/ps_in_mb)
+                       end if
```

Comment:
Some of these conditionals are ripe for the never-elser approach.  

You've at least got 2 levels of conditionals inside a do loop inside a select case inside a conditional, and ... I did not scroll up further than that.     I hate to slow the PR down, but I think we will do our future selves a bit of a favor with some refactoring here.  Please think about it.

Reply:
removed